src/backend/gl/src/window/egl.rs

```diff
@@ -38,35 +39,58 @@ fn get_x11_open_display() -> Option<ptr::NonNull<raw::c_void>> {
 }
 
 impl hal::Instance<crate::Backend> for Instance {
-    fn create(_: &str, _: u32) -> Result<Self, hal::UnsupportedBackend> {
+    fn create(_: &str, ptr: u32) -> Result<Self, hal::UnsupportedBackend> {
         let client_extensions = egl::query_string(None, egl::EXTENSIONS)
             .map_err(|_| hal::UnsupportedBackend)?
             .to_string_lossy();
         log::info!("Client extensions: {:?}", client_extensions);
         let client_ext_list = client_extensions.split_whitespace().collect::<Vec<_>>();
 
+        let mut is_x11 = false;
         let mut is_default_display = false;
         let x11_display = if client_ext_list.contains(&"EGL_EXT_platform_x11") {
             get_x11_open_display()
         } else {
             None
         };
-        let display = if let Some(x11_display) = x11_display {
-            log::info!("Using X11 platform");
-            const EGL_PLATFORM_X11_KHR: u32 = 0x31D5;
-            let display_attributes = [egl::ATTRIB_NONE];
-            egl::get_platform_display(
-                EGL_PLATFORM_X11_KHR,
-                x11_display.as_ptr(),
-                &display_attributes,
-            )
-            .unwrap()
+        let wl_display = if client_ext_list.contains(&"EGL_EXT_platform_wayland") {
+            wayland::display_connect()
         } else {
-            log::info!("Using default platform");
-            is_default_display = true;
-            egl::get_display(egl::DEFAULT_DISPLAY).unwrap()
+            None
+        };
+
+        let display = match (wl_display, x11_display) {
+            (Some(wl_display), _) => {
+                // Use egl::DEFAULT_DISPLAY so we can cleanup the connection
+                // we opened.
+                wayland::display_disconnect(wl_display.as_ptr() as _);
+                log::info!("Using Wayland platform");
+                const EGL_PLATFORM_WAYLAND_KHR: u32 = 0x31D8;
+                let display_attributes = [egl::ATTRIB_NONE];
+                egl::get_platform_display(EGL_PLATFORM_WAYLAND_KHR, ptr as _, &display_attributes)
```

Comment:
I don't understand this part. The docs on `EGL_PLATFORM_WAYLAND_KHR` say that a `wl_display` is needed here for the second parameter. And we have that right here, but instead of using it the code closes the connection and passes in the pointer that was given to `Instance::new`.

We shouldn't be modifying the signature of `Instance::new`.

Reply:
The problem is the `display` that is passed to `get_platform_display` is created by winit. If I create my own, then quad does not render at all. I understand modifying `Instance::new` here is undesired - I just did it prove it works. What I need is a handle to `wl_display` here that was created by winit.

Reply:
Would it be a different `wl_display`, necessarily, than what `wayland::display_connect()` returns?

Reply:
I'm not sure what is happening, but if I pass in `wl_display` that is returned from `wayland::display_connect()` then no window is created on screen. Same for passing `egl::DEFAULT_DISPLAY`.

Reply:
Sorry, how come no window is created? The window is created by winit, it has to be there.
Do you mean that no rendering is done on it? And no errors occurring here?